### PR TITLE
[Frontend] add tok/s speed metric to llm class when using tqdm

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
This PR adds a speed metric to the progress bar when using tqdm

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -233,22 +233,29 @@ def _run_engine(self, use_tqdm: bool) -> List[RequestOutput]:
...
         if use_tqdm:
             num_requests = self.llm_engine.get_num_unfinished_requests()
-            pbar = tqdm(total=num_requests,
-                        desc="Processed prompts",
-                        dynamic_ncols=True)
+            pbar = tqdm(
+                total=num_requests,
+                desc="Processed prompts",
+                dynamic_ncols=True,
+                postfix=f"{0} toks/s",
```
Comment (mgoin) [inline vllm/entrypoints/llm.py:240]: Just to make the starting postfix match formatting
```suggestion
                postfix=f"{0:.2f} toks/s",
```

**Comment by DamianB-BitFlipper** [comment]
This contribution would be much appreciated!

**Comment by MahmoudAshraf97** [comment]
@mgoin Sorry to bother you, I guess the failures in the CI aren't related to my change

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -233,22 +233,29 @@ def _run_engine(self, use_tqdm: bool) -> List[RequestOutput]:
...
         # Run the engine.
         outputs: List[RequestOutput] = []
+        total_tokens = 0
         while self.llm_engine.has_unfinished_requests():
             step_outputs = self.llm_engine.step()
             for output in step_outputs:
                 if output.finished:
                     outputs.append(output)
                     if use_tqdm:
+                        total_tokens += len(output.outputs[0].token_ids)
```
Comment (simon-mo) [inline vllm/entrypoints/llm.py:251]: This is not correct. We do see commonly multiple outputs due to setting `n>1`.

File: vllm/entrypoints/llm.py [hunk 3]
```diff
@@ -233,22 +233,29 @@ def _run_engine(self, use_tqdm: bool) -> List[RequestOutput]:
...
+                        speed = total_tokens / pbar.format_dict["elapsed"]
+                        pbar.postfix = f"{speed:.2f} toks/s"
```
Comment (simon-mo) [inline vllm/entrypoints/llm.py:253]: We should clarify this as decode/output/generation/generated tokens/s. This metrics does not capture the long prompt processing.

File: vllm/entrypoints/llm.py (hunk 2 above)
Comment (MahmoudAshraf97) [reply vllm/entrypoints/llm.py:251]: so the correct calculation is to sum over all outputs?

**Comment by simon-mo** [comment]
This is erroring in CI

https://buildkite.com/vllm/ci/builds/6751#018f529b-4a51-431a-b1d0-0a4daeb8d49f/319-360

```
35%|███▌      | 7/20 [00:04<00:08,  1.55it/s]--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.10/logging/__init__.py", line 1100, in emit
    msg = self.format(record)
  File "/usr/lib/python3.10/logging/__init__.py", line 943, in format
    return fmt.format(record)
  File "/usr/local/lib/python3.10/dist-packages/vllm/logger.py", line 24, in format
    msg = logging.Formatter.format(self, record)
  File "/usr/lib/python3.10/logging/__init__.py", line 678, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.10/logging/__init__.py", line 368, in getMessage
    msg = msg % self.args
ValueError: unsupported format character ',' (0x2c) at index 159
Call stack:
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.10/dist-packages/vllm/entrypoints/openai/api_server.py", line 169, in <module>
    uvicorn.run(app,
  File "/usr/local/lib/python3.10/dist-packages/uvicorn/main.py", line 575, in run
    server.run()
  File "/usr/local/lib/python3.10/dist-packages/uvicorn/server.py", line 65, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 470, in engine_step
    request_outputs = await self.engine.step_async()
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 226, in step_async
    self.stat_logger.log(self._get_stats(scheduler_outputs))
  File "/usr/local/lib/python3.10/dist-packages/vllm/engine/metrics.py", line 229, in log
    logger.info(
Message: 'Avg prompt throughput: %.1f tokens/s, Avg generation throughput: %.1f tokens/s, Running: %d reqs, Swapped: %d reqs, Pending: %d reqs, GPU KV cache usage: %.1f%, CPU KV cache usage: %.1f%'
Arguments: (154.2963163763204, 213.39696198061793, 13, 0, 0, 23.515151515151512, 0.0)
```

**Comment by MahmoudAshraf97** [comment]
@simon-mo I don't think this is related to my PR, it was fixed in https://github.com/vllm-project/vllm/pull/4396
I synced my fork and hopefully it will pass the CI

**Comment by simon-mo** [comment]
```



Processed prompts: 100%\|██████████\| 1000/1000 [01:02<00:00, 16.10it/s, Generation Speed: 4122.04 toks/s]
 

```

LGTM!

**Pull request merged**